Fix Dispatching not considering List[Optional[Tensor]] for dispatch

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60787 Fix Dispatching not considering List[Optional[Tensor]] for dispatch**

Fixes #60461.

Previously, when one calls `self.index(indices)` using a regular `self`
Tensor and a `BatchedTensor` indices the dispatcher would not dispatch
to the Batched key. This is because the dispatcher did not extract
dispatch keys from `indices`.

Similar #58283 and #58296, this PR modifies the dispatcher to extract
dispatch keys from List[Optional[Tensor]] arguments. We do this for both
boxed and unboxed kernels.

Test Plan:
- run the test case in
https://gist.github.com/zou3519/4421df7c5271376a0ef53ca857b18740
(requires functorch). After this PR, it raises `RuntimeError: Batching
rule not implemented for aten::index.Tensor. We could not generate a
fallback.`, which shows that dispatch happened on the Batched key.
- Taking suggestions for how to write a test for this in core

Differential Revision: [D29438611](https://our.internmc.facebook.com/intern/diff/D29438611)

zou3519 committed:
Fix Dispatching not considering List[Optional[Tensor]] for dispatch

Fixes #60461.

Previously, when one calls `self.index(indices)` using a regular `self`
Tensor and a `BatchedTensor` indices the dispatcher would not dispatch
to the Batched key. This is because the dispatcher did not extract
dispatch keys from `indices`.

Similar #58283 and #58296, this PR modifies the dispatcher to extract
dispatch keys from List[Optional[Tensor]] arguments. We do this for both
boxed and unboxed kernels.

Test Plan:
- run the test case in
https://gist.github.com/zou3519/4421df7c5271376a0ef53ca857b18740
(requires functorch). After this PR, it raises `RuntimeError: Batching
rule not implemented for aten::index.Tensor. We could not generate a
fallback.`, which shows that dispatch happened on the Batched key.
- Taking suggestions for how to write a test for this in core

[ghstack-poisoned]

bhosmer reviewed (approved):
LGTM! 

Re how best to add a test to core, that's a good question. I don't know of a place where we systematically test the coverage of this logic (hence the gaps we've been finding). Probably the most closely related tests are in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/op_registration/op_registration_test.cpp if you're down, but I wouldn't hold up landing this in any case.

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29438611).

zou3519 committed:
Update on "Fix Dispatching not considering List[Optional[Tensor]] for dispatch"


Fixes #60461.

Previously, when one calls `self.index(indices)` using a regular `self`
Tensor and a `BatchedTensor` indices the dispatcher would not dispatch
to the Batched key. This is because the dispatcher did not extract
dispatch keys from `indices`.

Similar #58283 and #58296, this PR modifies the dispatcher to extract
dispatch keys from List[Optional[Tensor]] arguments. We do this for both
boxed and unboxed kernels.

Test Plan:
- run the test case in
https://gist.github.com/zou3519/4421df7c5271376a0ef53ca857b18740
(requires functorch). After this PR, it raises `RuntimeError: Batching
rule not implemented for aten::index.Tensor. We could not generate a
fallback.`, which shows that dispatch happened on the Batched key.
- Taking suggestions for how to write a test for this in core

Differential Revision: [D29438611](https://our.internmc.facebook.com/intern/diff/D29438611)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29438611).
